### DISCUSSION TITLE
Write operations in DAG folder

### DISCUSSION THREAD

**Discussion by kmehkeri** [apache/airflow#40231] [General]
I run Airflow in AKS and have Azure Files share mounted read-only as DAG folder. I noticed that there's constant flow (like a dozen per second) of FileOpen operations for files named:

```
<dir>\__init__.so
<dir>\__init__.abi3.so
<dir>\__init__.cpython-310-x86_64-linux-gnu.so
```

in a couple of subdirectories (not all - maybe there's something specific in particular DAG files that triggers it - but I'm unable to isolate it). As the mount is read-only these operations fail, however Airflow operates normally and there's no trace of anything related to it in logs. I'm just billed for that and it's visible in storage account logs.

I read about Pyton shared object files, but I don't know where to go further. Any idea where these come from specifically? Is it Airflow itself or any providers or DAG code? Can these files be disabled or at least controlled where they are located similar to PYTHONPYCACHEPREFIX for pyc files?

**Comment by potiuk** [comment]
I believethis is controlled by `PYTHONBYTECODEBASE` variable: https://peps.python.org/pep-0304/ introduced in Python 3.8 and those files are generated following https://peps.python.org/pep-3147/ and https://peps.python.org/pep-3149/ that introduced per-ABI caches.

**Comment by kmehkeri** [comment]
Thanks for the hints. The critical part was here in [PEP 3149](https://peps.python.org/pep-3149/):

> Thus, an initial implementation of [PEP 384](https://peps.python.org/pep-0384/), when Python is configured with the default set of flags, would search for the following file names when extension module foo is imported (in this order):
> 
> ```
> foo.cpython-XYm.so
> foo.abi3.so
> foo.so
> ```

What I'm observing is Python checking whether those .so files exist (by attempting to read them). And it's recorded this way by Azure Files due to SMB protocol peculiarities (technically even for read there's a file handle created on server side that's interpreted as write).

I have a followup question though. What can be the reason the number of filesystem calls is so huge? My understanding would be that DAGs should be read by scheduler roughly once every `min_file_process_interval` / `dag_dir_list_interval` seconds. However this happens multiple times per second. With relatively small installation (we have minimal cluster / a few dozens of DAGs / a few hundred DAG runs daily) DAG files are being read few hundred times per hour and directories being open few thousand times per hour.

I noticed a similar question in StackOverflow [here](https://stackoverflow.com/questions/77648930/why-is-airflow-scheduler-making-so-many-system-calls-to-the-dag-files), but without any answers or comments. I think it might be the same issue.

**Reply by potiuk** [reply]
Possibly it might come from multiple tasks attempting to execute. The thing is that it's not only scheduler that parses the Python files, but also every task execution causes the DAG file to be parsed. So basically every time you run a task, the dag file that contains it is parsed by the "local task" run on the worker. So if you happen to have a lot of those tasks run in parallel, this might cause those multiple-per-second calls.

This is one more reason where I recommend - for big installation - rather than using shared filesystem to share the files, to use Git-sync approach, where each of the components involved have their own copy of the DAG folder in a local filesystem and any parsing there does not involve extra shared volume calls (it was not the primary reason I saw to avoid it, but it's quite similar to the problems I described in the article I wrote quite a long time ago - https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca - this might be yet another part of the "ugly" that I have not seen when I wrote it.

**Reply by kmehkeri** [reply]
The thing is my installation is quite small. It's just a few dozens of DAGs. We have a few up to max few hundred task instances per day. Filesystem access occurs constantly though, even when there's nothing at all running. The author of SO question I linked also straced scheduler process to prove that it was alone responsible for abundant calls. (Interestingly, their DAG storage was AWS mounted via NFS, mine is Azure Files via SMB).

Anyway, thanks for the hints again. I might also try to change my DAG storage/deployment model.

Or if I have some time I might try to investigate deeper what's going on in the scheduler myself. I suppose [dag_processing](https://github.com/apache/airflow/tree/main/airflow/dag_processing) would be a good place to start?

**Reply by potiuk** [reply]
Interesting. if anything happens in DAG file processor, you'd see logs generated in scheduler's logs. Another option is to start separately DAG file processor in standalone mode and see if it triggers the same kinds of issues.
